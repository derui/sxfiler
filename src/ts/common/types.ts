import * as fs from 'fs';

import {KeyDescriptor} from 'sxfiler-kbd';

export {KeyDescriptor} from 'sxfiler-kbd';

// the information for file
export interface File {
  filename: string;
  id: number;
  stat: fs.Stats;
}

export type KeyIdentifier = string;

// define mapping key to generics
export interface KeyMapping<T> {
  [key: string]: T;
}

// an interface for key bindings in handler
export interface KeyBindings<T> {
  bindings: KeyDescriptor[];
  mapping: KeyMapping<T>;
}
