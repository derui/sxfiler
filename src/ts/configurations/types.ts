import { ObjectEnum } from "@/utils";

export type ItemKey<T> = readonly [string, string, string];
export type SectionKey = readonly [string, string];
export type CategoryKey = readonly [string];

export type Descriptor<T> = {
  readonly key: T;
  readonly displayName: string;
  readonly description: string;
};

export type CategoryDescriptor = Descriptor<CategoryKey>;

export type SectionDescriptor = Descriptor<SectionKey>;

export type ItemDescriptor = Descriptor<ItemKey<any>>;

export type SelectOption = {
  value: any;
  toString(): string;
  display: string;
};

export const ItemElementType = {
  TEXT: "TEXT",
  NUMBER: "NUMBER",
  BOOLEAN: "BOOLEAN",
  SELECT_ONE: "SELECT_ONE",
} as const;
export type ItemElementType = ObjectEnum<typeof ItemElementType>;

export type ItemKind<K extends ItemElementType> = {
  kind: K;
};

export type TextItemType = ItemKind<"TEXT"> & {
  defaultValue?: string;
};
export type NumberItemType = ItemKind<"NUMBER"> & {
  defaultValue?: number;
};
export type BooleanItemType = ItemKind<"BOOLEAN"> & {
  defaultValue: boolean;
};
export type SelectOneItemType = ItemKind<"SELECT_ONE"> & {
  options: SelectOption[];
  defaultValue: string;
};

export type ItemType = TextItemType | NumberItemType | BooleanItemType | SelectOneItemType;

export type Item = ItemDescriptor & {
  readonly type: ItemType;
};

export type Section = SectionDescriptor & {
  readonly items: Item[];
};

export type Category = CategoryDescriptor & {
  readonly sections: Section[];
};
