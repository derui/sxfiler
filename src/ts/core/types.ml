(** {!Mode} defines type that is current mode of view. *)
module Mode = struct
  type t =
    | File_tree
    | Preview
    | Complete

  let all_modes = [File_tree; Preview; Complete]

  let to_context = function
    | File_tree -> "onFileTree"
    | Preview -> "onPreview"
    | Complete -> "onComplete"
end

module File_list_pos = struct
  type t =
    [ `Left
    | `Right ]

  let to_string = function `Left -> "left" | `Right -> "right"
end

type corrections = Sxfiler_rpc.Types.Node.t list
