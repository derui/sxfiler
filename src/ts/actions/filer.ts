import { AppAction, ActionTypes } from "./type";
import { Filer } from "@/domains/filer";
import { Side } from "@/states/file-list";

type SelectAction = AppAction<ActionTypes.FILER_SELECT, { side: Side; itemId: string }>;
type ReloadAction = AppAction<
  ActionTypes.FILER_RELOAD,
  {
    payload: {
      filers: [Filer, Filer];
    };
  }
>;

type UpdateFilerAction = AppAction<
  ActionTypes.FILER_UPDATE,
  {
    payload: {
      filer: Filer;
    };
  }
>;

// The action to load or get a filer from server
type LoadFilerAction = AppAction<
  ActionTypes.FILER_LOAD,
  {
    payload: {
      filer: Filer;
    };
  }
>;

type ChangeSideAction = AppAction<ActionTypes.FILER_CHANGE_SIDE>;

export type Actions = ReloadAction | UpdateFilerAction | ChangeSideAction | LoadFilerAction | SelectAction;

export const reload = function reload(args: { filers: [Filer, Filer] }): ReloadAction {
  return { type: ActionTypes.FILER_RELOAD, payload: { ...args } };
};

/**
 * Update filer
 */
export const update = function update(args: { filer: Filer }): UpdateFilerAction {
  return { type: ActionTypes.FILER_UPDATE, payload: { ...args } };
};

/**
 * Load or get filer on the side.
 */
export const load = function load(args: { filer: Filer }): LoadFilerAction {
  return { type: ActionTypes.FILER_LOAD, payload: { ...args } };
};

/**
   change current side
*/
export const changeSide = function changeSide(): ChangeSideAction {
  return { type: ActionTypes.FILER_CHANGE_SIDE };
};

export const select = function select(side: Side, itemId: string): SelectAction {
  return { type: ActionTypes.FILER_SELECT, side, itemId };
};
