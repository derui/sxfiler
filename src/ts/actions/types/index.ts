// all actions in this application
import NotificationActionTypes from "./notification";

type ActionTypes = NotificationActionTypes;

export default ActionTypes;
