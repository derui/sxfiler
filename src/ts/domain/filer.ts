// define filer state type and operations.
import { Node } from "./node";

export interface NodeMarker {
  node: Node;
  marked: boolean;
}

export enum Direction {
  Up = "up",
  Down = "down",
}

// Filer domain to handle nodes in the location
export class Filer {
  constructor(
    public readonly id: string,
    public readonly location: string,
    public readonly nodes: NodeMarker[],
    public readonly currentCursorIndex: number
  ) {}

  get currentNode(): Node {
    return this.nodes[this.currentCursorIndex].node;
  }

  /**
   * Get new instance of Filer that is moved by direction
   * @param direction direction to move index
   */
  public moveIndex(direction: Direction): Filer {
    let index = 0;

    switch (direction) {
      case Direction.Down:
        index = Math.min(this.nodes.length - 1, this.currentCursorIndex + 1);
        break;
      case Direction.Up:
        index = Math.max(0, this.currentCursorIndex - 1);
        break;
    }

    return new Filer(this.id, this.location, this.nodes, index);
  }

  /**
   * Mark specified node
   * @param nodeId
   */
  public toggleMark(nodeId: string): Filer {
    return new Filer(
      this.id,
      this.location,
      this.nodes.map(v => {
        if (v.node.id === nodeId) {
          return { node: v.node, marked: !v.marked };
        }
        return v;
      }),
      this.currentCursorIndex
    );
  }

  /**
   * @returns marked nodes
   */
  get markedNodes(): Node[] {
    return this.nodes.filter(v => v.marked).map(v => v.node);
  }
}
