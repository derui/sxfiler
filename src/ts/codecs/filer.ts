import { Filer, createFiler } from "../domains/filer";
import { FileItemOnRPC, encode as encodeFileItem } from "./file-item";

// define codec that is between filer domain and RPC

export type FilerOnRPC = {
  id: string;
  name: string;
  fileList: {
    location: string;
    items: FileItemOnRPC[];
  };
  markedItems: string[];
  sortOrder: string;
};

/**
   encode filer object from RPC to frontend domain.

   @param filer JSON representation for fller
   @return Filer object
 */
export const encode = (obj: FilerOnRPC): Filer => {
  const {
    id,
    name,
    fileList: { location, items },
    markedItems,
  } = obj;

  return createFiler({
    id,
    name,
    location,
    items: items
      .map(v => ({
        ...v,
        marked: markedItems.includes(v.id),
      }))
      .map(encodeFileItem),
    currentCursorIndex: 0,
  });
};
