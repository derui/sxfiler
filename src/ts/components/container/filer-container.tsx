import { h } from "preact";
import * as FileList from "@/components/project/file-list";
import * as F from "@/modules/filer";
import * as FilerReducer from "@/modules/filer/reducer";
import * as FilerPb from "@/generated/filer_pb";
import { Type } from "@/types/natural-number";
import {
  leftSideCursorPositionSelector,
  leftSideFileListSelector,
  rightSideFileListSelector,
  rightSideCursorPositionSelector,
} from "@/modules/filer/selectors";

export type Props = {
  state: F.State;
};

/* create filer from state and key */
const createFiler = (
  side: FilerReducer.Side,
  currentCursorIndex: Type,
  currentSide: FilerReducer.Side,
  fileList?: FilerPb.FileList
): FileList.ElementType | null => {
  const focused = side === currentSide;

  if (!fileList) {
    return null;
  }

  return (
    <FileList.Component
      key={side}
      items={fileList.toObject().itemsList}
      cursor={currentCursorIndex.value}
      location={fileList.toObject().location}
      focused={focused}
      bookmarks={[]}
    />
  );
};

// Stateless container to render filer
export const Component: preact.FunctionComponent<Props> = ({ state }): ElementType | null => {
  // can not render anything if filer is not initialized
  const filer = state.filer;
  if (!filer) {
    return null;
  }

  const leftFiler = createFiler(
    FilerReducer.Side.Left,
    leftSideCursorPositionSelector(state),
    state.currentSide,
    leftSideFileListSelector(state)
  );
  const rightFiler = createFiler(
    FilerReducer.Side.Right,
    rightSideCursorPositionSelector(state),
    state.currentSide,
    rightSideFileListSelector(state)
  );

  return (
    <div class="filer-container__root" data-testid="filerContainer">
      {leftFiler}
      <div class="filer-container__separator" data-testid="filerContainer-separator" />
      {rightFiler}
    </div>
  );
};

export type ElementType = ReturnType<typeof Component>;
