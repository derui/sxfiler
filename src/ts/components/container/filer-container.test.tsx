import { h } from "preact";
import { render } from "preact-render-to-string";

import { Component } from "./filer-container";
import { also } from "@/libs/fn";
import { Filer, FileWindow, FileList } from "@/generated/filer_pb";
import { State, Side } from "@/modules/filer/reducer";
import * as N from "@/types/natural-number";

describe("Container", () => {
  let state: State = {
    currentSide: Side.Left,
    currentCursorPosition: {
      left: N.create(0),
      right: N.create(0),
    },
    filer: also(new Filer(), (v) => {
      v.setLeftFileWindow(
        also(new FileWindow(), (v) => {
          v.setFileList(
            also(new FileList(), (v) => {
              v.setLocation("/left");
              v.setItemsList([]);
            })
          );
        })
      );
    }),
  };

  describe("File List Container", () => {
    it("should render correctly", () => {
      const tree = render(<Component state={state} />);

      expect(tree).toMatchSnapshot();
    });
  });
});
