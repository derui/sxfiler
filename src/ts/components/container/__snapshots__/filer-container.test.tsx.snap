// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container File List Container should render correctly 1`] = `"<div class=\\"filer-container__root\\" data-testid=\\"filerContainer\\"><div class=\\"file-list__root\\" data-testid=\\"fileList-root\\"><header class=\\"file-list__header\\" data-testid=\\"fileList-header\\" data-focused>/left</header><div style=\\"overflow: visible;\\" class=\\"file-list__autosizer\\"></div></div><div class=\\"filer-container__separator\\" data-testid=\\"filerContainer-separator\\"></div></div>"`;
